src/test/groovy/edu/ie3/simona/model/participant/StorageModelTest.groovy

```diff
@@ -0,0 +1,374 @@
+    OperationTime.notLimited(),
+    getQuantity(1d, PU),
+    false,
+    NodeInput.DEFAULT_GEO_POSITION,
+    GermanVoltageLevelUtils.LV,
+    -1)
+
+    def typeInput = new StorageTypeInput(
+    UUID.fromString("fbee4995-24dd-45e4-9c85-7d986fe99ff3"),
+    "Test_StorageTypeInput",
+    getQuantity(100d, EURO),
+    getQuantity(101d, EURO_PER_MEGAWATTHOUR),
+    getQuantity(100d, KILOWATTHOUR),
+    getQuantity(13d, KILOVOLTAMPERE),
+    0.997,
+    getQuantity(10d, KILOWATT),
+    getQuantity(0.03, PU_PER_HOUR),
+    getQuantity(0.9, PU),
+    )
+
+    inputModel = new StorageInput(
+    UUID.randomUUID(),
+    "Test_StorageInput",
+    new OperatorInput(UUID.randomUUID(), "NO_OPERATOR"),
+    OperationTime.notLimited(),
+    nodeInput,
+    CosPhiFixed.CONSTANT_CHARACTERISTIC,
+    null,
+    typeInput
+    )
+  }
+
+  def buildStorageModel(Option<Double> targetSoc = Option.empty()) {
+    return StorageModel.apply(inputModel, 1,
+    TimeUtil.withDefaults.toZonedDateTime("2020-01-01T00:00:00Z"),
+    TimeUtil.withDefaults.toZonedDateTime("2020-01-01T01:00:00Z"),
+    0d,
+    targetSoc)
+  }
+
+  def "Calculate flex options"() {
+    given:
+    def storageModel = buildStorageModel()
+    def startTick = 3600L
+    def data = new StorageModel.StorageRelevantData(startTick + timeDelta)
+    def oldState = new StorageModel.StorageState(
+    Sq.create(lastStored.doubleValue(), KilowattHours$.MODULE$),
+    Sq.create(lastPower.doubleValue(), Kilowatts$.MODULE$),
+    startTick
+    )
+
+    when:
+    def result = (ProvideMinMaxFlexOptions) storageModel.determineFlexOptions(data, oldState)
+
+    then:
+    Math.abs(result.ref().toKilowatts() - pRef) < TOLERANCE
+    Math.abs(result.min().toKilowatts() - pMin) < TOLERANCE
+    Math.abs(result.max().toKilowatts() - pMax) < TOLERANCE
+
+    where:
+    lastStored   | lastPower | timeDelta || pRef | pMin | pMax
+    // UNCHANGED STATE
+    // completely empty
+    0            | 0         | 1         || 0    | 0    | 10
+    // at a tiny bit above empty
+    0.011d       | 0         | 1         || 0    | -10  | 10
+    // at mid-level charge
+    60           | 0         | 1         || 0    | -10  | 10
+    // almost fully charged
+    99.989d      | 0         | 1         || 0    | -10  | 10
+    // fully charged
+    100          | 0         | 1         || 0    | -10  | 0
+    // CHANGED STATE
+    // discharged to empty
+    10           | -9        | 3600      || 0    | 0    | 10
+    // almost discharged to lowest allowed charge
+    10           | -9        | 3590      || 0    | -10  | 10
+    // charged to mid-level charge
+    41           | 10        | 3600      || 0    | -10  | 10
+    // discharged to mid-level charge
+    60           | -9        | 3600      || 0    | -10  | 10
+    // almost fully charged
+    95.5         | 4.98      | 3600      || 0    | -10  | 10
+    // fully charged
+    95.5         | 5         | 3600      || 0    | -10  | 0
```

Comment:
This is fully charged because with 5 kW over one hour it will be above 100%, right?

Reply:
No, we're exactly fully charged because with 90% efficiency, we're net charging with 4.5 kW

Reply:
Got it, thanks!